### PR TITLE
[1/X] Add algorithm to schedule new meeting.

### PR DESCRIPTION
**Pull request by nikhitashanker**
Algorithm is as follow:
- Find potential conflicts by using attendees list of request and currently scheduled events.
- Sort the potential conflicts by ascending order of start time.
- Iterate through the conflicts, keeping track of latest end time seen so far.
    - If the time between the end time of the current conflict and the latest end time is 
       a valid time range, add it to the returned time ranges.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -13,11 +13,60 @@
 // limitations under the License.
 
 package com.google.sps;
-
+import java.util.Collections;
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.HashSet;
```
Comment (yicyao) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:21]: alphabetize imports

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -13,11 +13,60 @@
...
+    Collection<TimeRange> timeRanges = new ArrayList<TimeRange>();
+    int lastConflictEnd = TimeRange.START_OF_DAY;
+    long requestDuration = request.getDuration();
+    for (TimeRange currentConflict : potentialConflicts) {
+        // If the time between the end time of the current conflict and the latest end time 
+        // conflict seen so far is greater that the request duration provided,
+        // add a new time range to the time ranges returned.
+        int currentConflictStart = currentConflict.start();
+        if (currentConflictStart - lastConflictEnd >= requestDuration) {
+            timeRanges.add(TimeRange.fromStartEnd(lastConflictEnd, currentConflictStart, false));
```
Comment (yicyao) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:41]: can add what 'false' is eg`  timeRanges.add(TimeRange.fromStartEnd(lastConflictEnd, currentConflictStart, *exclusive interval */ false));`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (nikhitashanker) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (nikhitashanker) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done. Good idea - thanks.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,60 @@
...
+    }
+    return timeRanges;
+  }
+
+  private static List<TimeRange> getPotentialConflicts(Collection<Event> events, MeetingRequest request) {
+    List<TimeRange> potentialConflicts = new ArrayList<TimeRange>();
+    Collection<String> requestAttendees = request.getAttendees();
+    for (Event event : events) {
+        // Find the intersection between current event attendees and request attendees.
+        HashSet<String> intersection = new HashSet<String>(event.getAttendees());
```
Comment (dash102) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:63]: I think usually you'd declare a HashSet (or LinkedHashSet or whatever else implements Set) as a Set to keep it as generic as possible

Comment (nikhitashanker) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done. Noted.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,60 @@
...
+        // conflict seen so far is greater that the request duration provided,
+        // add a new time range to the time ranges returned.
+        int currentConflictStart = currentConflict.start();
+        if (currentConflictStart - lastConflictEnd >= requestDuration) {
+            timeRanges.add(TimeRange.fromStartEnd(lastConflictEnd, currentConflictStart, /* inclusive= */ false));
+        }
+
+        // Update the latest conflict end time seen so far.
+        int currentConflictEnd = currentConflict.end();
+        if (currentConflictEnd > lastConflictEnd)
```
Comment (hlz03) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:47]: format: if () {
};

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,60 @@
...
+    List<TimeRange> potentialConflicts = new ArrayList<TimeRange>();
+    Collection<String> requestAttendees = request.getAttendees();
+    for (Event event : events) {
+        // Find the intersection between current event attendees and request attendees.
+        Set<String> intersection = new HashSet<String>(event.getAttendees());
+        intersection.retainAll(requestAttendees);
+
+        // If the intersection size is greater than 0, add this event time to the potential
+        // conflicts.
+        if (intersection.size() > 0)
```
Comment (hlz03) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:68]: intersection.isEmpty()

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (nikhitashanker) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:47]: Done.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (nikhitashanker) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:68]: Done.

**Pull request merged**